nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py

```diff
@@ -0,0 +1,493 @@
+            answer_ids = answer_ids[: -min(truncation_length, len(answer_ids))]
+            context_ids = context_ids[: -min(truncation_length, len(context_ids))]
+
+        input_ids = context_ids
+        answer_start_idx = len(input_ids)
+
+        # Adds bos token in the start
+        if self.add_bos:
+            context_ids = [self.bos_id] + context_ids
+            input_ids = [self.bos_id] + input_ids
+            answer_start_idx += 1
+
+        # Adds sep token between text/prompt and answer
+        if self.add_sep:
+            context_ids = context_ids + [self.sep_id]
+            input_ids = input_ids + [self.sep_id]
+            answer_start_idx += 1
+
+        # create a copy of answer_ids and mask on it
+        if self.input_text_mask_ratio is not None and self.input_text_mask_ratio > 0:
+            if self.sample_alpha is None:
+                masked_answer_ids, _ = self._random_mask_tokens(
+                    answer_ids, self.input_text_mask_ratio, self.tokenizer.unk_id
+                )
+            else:
+                sample_answer_ids = pre_pad + _text_to_ids(answer_text, self.sample_alpha, lang=lang)
+                # does not consider different length for now
+                masked_answer_ids, _ = self._random_mask_tokens(
+                    answer_ids,
+                    self.input_text_mask_ratio,
+                    self.tokenizer.unk_id,
+                    sample_tokens=sample_answer_ids,
+                )
+            masked_input_ids = input_ids + masked_answer_ids
+        input_ids = input_ids + answer_ids
+
+        # Only training need to consider eos token
+        if self.add_eos:
+            input_ids = input_ids + [self.tokenizer.eos_id]
+            answer_ids = answer_ids + [self.tokenizer.eos_id]
+            if self.input_text_mask_ratio is not None and self.input_text_mask_ratio > 0:
+                masked_input_ids = masked_input_ids + [self.tokenizer.eos_id]
+
+        if len(input_ids) > self.max_seq_length:
+            logging.warning(f'Input ids length {len(input_ids)} exceed max sequence length {self.max_seq_length}')
+            input_ids = input_ids[: self.max_seq_length]
+            if self.input_text_mask_ratio is not None and self.input_text_mask_ratio > 0:
+                masked_input_ids = masked_input_ids[: self.max_seq_length]
+
+        processed_example = {
+            'input_ids': torch.as_tensor(input_ids),
+            'answer_start_idx': torch.as_tensor(answer_start_idx),
+            'context_ids': torch.as_tensor(context_ids),
+            'context_length': len(context_ids),
+            'answer_ids': torch.as_tensor(answer_ids),
+        }
+
+        if self.input_text_mask_ratio is not None and self.input_text_mask_ratio > 0:
+            processed_example['masked_input_ids'] = torch.as_tensor(masked_input_ids)
+
+        return processed_example
+
+
+def convert_canary_prompt_to_text(prompt, is_canary_tokens_augment):
```

Comment:
I understand why this function is built the way it is but for future experiments let's try to move away from canary special token conversion and design a configurable prompting setup instead.

Reply:
acknowledge. Add a TODO